applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_coupling_interface_condition.cpp

```diff
@@ -0,0 +1,238 @@
+                                                        CalculateResidualVectorFlag);
+
+    // Append penalty force at the nodes
+    if (Is(INTERFACE) && !mReactionIsAdded)
+    {
+        this->CalculateNodalContactForce( rRightHandSideVector, rCurrentProcessInfo, CalculateResidualVectorFlag );
+        mReactionIsAdded = true;
+    }
+
+    KRATOS_CATCH( "" )
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void MPMParticlePenaltyCouplingInterfaceCondition::CalculateNodalContactForce( const VectorType& rRightHandSideVector, ProcessInfo& rCurrentProcessInfo, const bool CalculateResidualVectorFlag )
+{
+    if ( CalculateResidualVectorFlag == true )
+    {
+        GeometryType& rGeom = GetGeometry();
+        const unsigned int number_of_nodes = rGeom.size();
+        const unsigned int dimension = rGeom.WorkingSpaceDimension();
+        const unsigned int block_size = this->GetBlockSize();
+
+        // Calculate nodal forces
+        Vector nodal_force = ZeroVector(3);
+        for (unsigned int i = 0; i < number_of_nodes; i++)
+        {
+            for (unsigned int j = 0; j < dimension; j++)
+            {
+                nodal_force[j] = rRightHandSideVector[block_size * i + j];
+            }
+
+            // Check whether there nodes are active and associated to material point elements
+            const double& nodal_mass = rGeom[i].FastGetSolutionStepValue(NODAL_MASS, 0);
+            if (nodal_mass > std::numeric_limits<double>::epsilon())
+            {
+                rGeom[i].SetLock();
+                rGeom[i].FastGetSolutionStepValue(REACTION) += nodal_force;
+                rGeom[i].UnSetLock();
+            }
+
+        }
+    }
+}
+
+//************************************************************************************
+//************************************************************************************
+
+void MPMParticlePenaltyCouplingInterfaceCondition::CalculateInterfaceContactForce( ProcessInfo& rCurrentProcessInfo )
+{
+    GeometryType& rGeom = GetGeometry();
+    const unsigned int number_of_nodes = rGeom.PointsNumber();
+    // const unsigned int dimension = rGeom.WorkingSpaceDimension();
+
+    // Prepare variables
+    GeneralVariables Variables;
+    const array_1d<double, 3 > & xg_c = this->GetValue(MPC_COORD);
+    const double & MPC_Area = this->GetValue(MPC_AREA);
+    Variables.N = this->MPMShapeFunctionPointValues(Variables.N, xg_c);
+
+    // Interpolate the force to MPC_Force assuming linear shape function
+    array_1d<double, 3 > MPC_Force = ZeroVector(3);
+    for (unsigned int i = 0; i < number_of_nodes; i++)
+    {
+        // Check whether there is material point inside the node
+        const double& nodal_mass = rGeom[i].FastGetSolutionStepValue(NODAL_MASS, 0);
+        const double nodal_area  = rGeom[i].FastGetSolutionStepValue(NODAL_AREA, 0);
```

Comment:
Integration Weight?

Should be integration weight * differential length of segment (DeterminantOfJacobian())

Reply:
This is the accumulative nodal_area computed through extrapolation from the particle integration weight. This is made to weight the contact force value evaluated at the nodes (see the details in my master thesis).

Reply:
I see. Ok I see, integration is done over the FEM. You can keep it. However, we should keep in mind, as soon as we have a Mortar, this can be neglected, as the value on the condition will be already including this integrational area.